albatross/evaluation/cross_validation_utils.h

```diff
@@ -0,0 +1,157 @@
+}
+
+template <typename FeatureType>
+inline Eigen::VectorXd concatenate_mean_predictions(
+    const std::vector<RegressionFold<FeatureType>> &folds,
+    const std::vector<Eigen::VectorXd> &means) {
+  assert(folds.size() == means.size());
+
+  Eigen::Index n = static_cast<Eigen::Index>(max_index(folds));
+  Eigen::VectorXd pred(n);
+  Eigen::Index number_filled = 0;
+  // Put all the predicted means back in order.
+  for (std::size_t i = 0; i < folds.size(); ++i) {
+    const auto fold_pred = means[i];
+    const auto fold = folds[i];
+    Eigen::Index fold_size =
+        static_cast<Eigen::Index>(fold.test_dataset.size());
+    assert(fold_pred.size() == fold_size);
+    set_subset(fold.test_indices, fold_pred, &pred);
+    number_filled += static_cast<Eigen::Index>(fold.test_indices.size());
+  }
+  assert(number_filled == n);
+  return pred;
+}
+
+template <typename FeatureType, typename PredType>
+inline Eigen::VectorXd concatenate_mean_predictions(
+    const std::vector<RegressionFold<FeatureType>> &folds,
+    const std::vector<PredType> &predictions) {
+  return concatenate_mean_predictions(folds, get_means(predictions));
+}
+
+template <typename FeatureType>
+inline MarginalDistribution concatenate_marginal_predictions(
+    const std::vector<RegressionFold<FeatureType>> &folds,
+    const std::vector<MarginalDistribution> &marginals) {
+  assert(folds.size() == marginals.size());
+
+  Eigen::Index n = static_cast<Eigen::Index>(max_index(folds));
+  assert(folds.size() == marginals.size());
+  Eigen::VectorXd mean(n);
+  Eigen::VectorXd variance(n);
+  Eigen::Index number_filled = 0;
+  // Put all the predicted means back in order.
+  for (std::size_t i = 0; i < folds.size(); ++i) {
+    const auto fold_pred = marginals[i];
+    const auto fold = folds[i];
+    assert(fold_pred.size() == fold.test_dataset.size());
+    set_subset(fold.test_indices, fold_pred.mean, &mean);
+    set_subset(fold.test_indices, fold_pred.covariance.diagonal(), &variance);
+    number_filled += static_cast<Eigen::Index>(fold.test_indices.size());
+  }
+  assert(number_filled == n);
+  return MarginalDistribution(mean, variance.asDiagonal());
+}
+
+template <typename FeatureType, typename PredType>
+inline MarginalDistribution concatenate_marginal_predictions(
+    const std::vector<RegressionFold<FeatureType>> &folds,
+    const std::vector<PredType> &predictions) {
+  return concatenate_marginal_predictions(folds, get_marginals(predictions));
+}
+
+template <typename FeatureType, typename RequiredPredictType>
+Eigen::VectorXd
+cross_validated_scores(const EvaluationMetric<RequiredPredictType> &metric,
+                       const std::vector<RegressionFold<FeatureType>> &folds,
+                       const std::vector<RequiredPredictType> &predictions) {
+  assert(folds.size() == predictions.size());
+  Eigen::Index n = static_cast<Eigen::Index>(predictions.size());
+  Eigen::VectorXd output(n);
+  for (Eigen::Index i = 0; i < n; ++i) {
```

Comment:
the output is an `Eigen::VectorXd` is that still compatible with `std::transform`?

Reply:
`std::transform` operates on iterator like objects. So assuming `Eigen::VectorXd` provides an interator object it would look something like this 

```
template <typename FeatureType, typename RequiredPredictType>
Eigen::VectorXd
cross_validated_scores(const EvaluationMetric<RequiredPredictType> &metric,
                       const std::vector<RegressionFold<FeatureType>> &folds,
                       const std::vector<RequiredPredictType> &predictions) {
  assert(folds.size() == predictions.size());
  Eigen::Index n = static_cast<Eigen::Index>(predictions.size());
  Eigen::VectorXd output(n);
  auto make_metric = [](const auto &fold, const auto& prediction) {
    auto fold_size = static_cast<std::size_t>(fold.test_dataset.size());
    auto pred_size = static_cast<std::size_t>(prediction.size());
    assert(fold_size == pred_size);
    return metric(prediction, fold.test_dataset.targets);
  }
  std::transform(folds.begin(), folds.end(), predictions.begin(), output.begin(), make_metric);
}
```

Like I said, probably doesn't increase readability much.

Reply:
Ah yeah, I was assuming `Eigen::VectorXd` didn't support iterators, but it does as of 3.4: `https://eigen.tuxfamily.org/dox-devel/group__TutorialSTL.html`